recipe-server/client/control_new/state/app/recipes/reducers.js

```diff
@@ -38,23 +38,31 @@ function history(state = new Map(), action) {
   }
 }
 
+const formatRecipe = recipe =>
+  recipe
+    .set('action_id', recipe.getIn(['action', 'id'], null))
+    .set('latest_revision_id', recipe.getIn(['latest_revision', 'id'], null))
+    .set('approved_revision_id', recipe.getIn(['approved_revision', 'id'], null))
+    .remove('action')
+    .remove('latest_revision')
+    .remove('approved_revision');
 
 function items(state = new Map(), action) {
-  let recipe;
-
   switch (action.type) {
     case RECIPE_RECEIVE: {
-      recipe = fromJS(action.recipe);
+      const recipe = fromJS(action.recipe);
+      return state.set(action.recipe.id, formatRecipe(recipe));
+    }
+
+    case RECIPE_PAGE_RECEIVE: {
+      const recipes = fromJS(action.recipes.results);
+      let newState = state;
 
-      recipe = recipe
-        .set('action_id', recipe.getIn(['action', 'id'], null))
-        .set('latest_revision_id', recipe.getIn(['latest_revision', 'id'], null))
-        .set('approved_revision_id', recipe.getIn(['approved_revision', 'id'], null))
-        .remove('action')
-        .remove('latest_revision')
-        .remove('approved_revision');
+      recipes.forEach(receivedRecipe => {
+        newState = newState.set(receivedRecipe.get('id'), formatRecipe(receivedRecipe));
+      });
```

Comment:
As a drive comment, Immutable's [`withMutations` helper](https://facebook.github.io/immutable-js/docs/#/Map/withMutations) might help speed this up. It temporarily makes the data structure directly mutable, so you could write this as:

```js
const recipes = fromJS(action.recipes.results);
return state.withMutations(mutState => {
  recipe.forEach(receivedRecipe => {
    mutState.set(receivedRecipe.get('id'), formatRecipe(receivedRecipe));
  });
});
```